src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,124 @@
+                    (arr) =>
+                    {
+                        return string.Join(',', arr);
+                    });
+
+            Assert.Throws<Exception>(() => _df.Select(udfInt(_df["ages"])).Collect().ToArray());
+
+            Func<Column, Column> udfString = Udf<string[], string>(
+                    (arr) =>
+                    {
+                        return string.Join(',', arr);
+                    });
+
+            Assert.Throws<Exception>(() => _df.Select(udfString(_df["cars"])).Collect().ToArray());
+        }
+
+        // UDF that returns Array type.
+        [Fact]
+        public void TestUdfWithReturnTypeAsArray()
+        {
+            Func<Column, Column> udf = Udf<string, string[]>(
+                    (str) =>
+                    {
+                        return new string[] { str, str + str };
+                    });
+
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        // UDF that takes in Map type.
+        [Fact]
+        public void TestUdfWithMapType()
+        {
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    (dict) =>
+                    {
+                        return dict.Count.ToString();
+                    });
+
+            Assert.Throws<Exception>(() => _df.Select(udf(_df["info"])).Collect().ToArray());
+        }
+
+        // UDF that returns Map type.
+        [Fact]
+        public void TestUdfWithReturnTypeAsMap()
+        {
+            Func<Column, Column> udf = Udf<string, IDictionary<string, string>>(
+                (str) =>
+                {
+                    return new Dictionary<string, string> { { str, str } };
+                });
+
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        // UDF that takes in Row type.
+        [Fact]
+        public void TestUdfWithRowType()
+        {
+            Func<Column, Column> udf = Udf<Row, string>(
+                (row) =>
+                {
+                    string city = row.GetAs<string>("city");
+                    string state = row.GetAs<string>("state");
+                    return $"{city},{state}";
+                });
+
+            Row[] rows = _df.Select(udf(_df["info"])).Collect().ToArray();
+            Assert.Equal(3, rows.Length);
+
+            var expected = new[] { "Burdwan,Paschimbanga", "Los Angeles,California", "Seattle," };
+            for (int i = 0; i < rows.Length; ++i)
+            {
+                Row row = rows[i];
+                Assert.Equal(1, row.Size());
+                Assert.Equal(expected[i], row.GetAs<string>(0));
+            }
+        }
+
+        // UDF that returns Row type.
+        [Fact]
+        public void TestUdfWithReturnTypeAsRow()
+        {
+            Assert.Throws<ArgumentException>(() => Udf<string, object[]>(
+                (str) =>

```

Comment:
1-liner

Reply:
I have updated accordingly to resolve all of the commits. Thanks!